File: frontend/src/services/flows.ts
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8000/api';

export interface Flow {
  id: number;
  name: string;
  description: string | null;
  serialized_graph: any;
  created_at?: string;
  updated_at?: string;
}

export interface CreateFlowData {
  name: string;
  description?: string;
  serialized_graph: any;
}

// Get all flows
export const getFlows = async (limit?: number): Promise<Flow[]> => {
  try {
    const response = await axios.get(`${API_BASE_URL}/flows/`, {
      params: { limit }
    });
    return response.data;
  } catch (error) {
    console.error('Error fetching flows:', error);
    throw error;
  }
};

// Get a single flow by ID
export const getFlow = async (id: number): Promise<Flow> => {
  try {
    const response = await axios.get(`${API_BASE_URL}/flows/${id}`);
    return response.data;
  } catch (error) {
    console.error(`Error fetching flow ${id}:`, error);
    throw error;
  }
};

// Create a new flow
export const createFlow = async (flowData: CreateFlowData): Promise<Flow> => {
  try {
    const response = await axios.post(`${API_BASE_URL}/flows/`, flowData);
    return response.data;
  } catch (error) {
    console.error('Error creating flow:', error);
    throw error;
  }
};

// Update an existing flow
export const updateFlow = async (id: number, flowData: CreateFlowData): Promise<Flow> => {
  try {
    const response = await axios.put(`${API_BASE_URL}/flows/${id}`, flowData);
    return response.data;
  } catch (error) {
    console.error(`Error updating flow ${id}:`, error);
    throw error;
  }
};

// Delete a flow
export const deleteFlow = async (id: number): Promise<Flow> => {
  try {
    const response = await axios.delete(`${API_BASE_URL}/flows/${id}`);
    return response.data;
  } catch (error) {
    console.error(`Error deleting flow ${id}:`, error);
    throw error;
  }
};

File: backend/api/flows.py
from fastapi import APIRouter, Depends, HTTPException
from typing import Optional
from sqlalchemy.orm import Session
from schemas.flows import FlowCreate, FlowOut, FlowPayload
from crud.flows import create_flow, get_flow_by_id, update_flow_by_id, delete_flow_by_id, get_flows
from db.session import get_db
from services.flows.codegen import CodeGenerator

router = APIRouter(
    prefix="/flows",
    tags=["Flow"],
    responses={404: {"description": "Not found"}},
)


#########################
## Simple Flow Operations
#########################

@router.post("/", description="Add a new flow")
def add_flow(flow: FlowCreate, db: Session = Depends(get_db)):
    return create_flow(db, name=flow.name, description=flow.description, graph=flow.graph, state=flow.state)


@router.get("/", description="List all flows")
def list_flows(limit: Optional[int] = None, db: Session = Depends(get_db)):
    return get_flows(db, limit)


@router.get("/{id}", description="Get a flow by ID", response_model=FlowOut)
def get_flow(id: int, db: Session = Depends(get_db)):
    return get_flow_by_id(db, id)


@router.put("/{id}", description="Update a flow by ID", response_model=FlowOut)
def update_flow(id: int, flow: FlowCreate, db: Session = Depends(get_db)):
    updated = update_flow_by_id(db, id, flow.name, flow.description, flow.graph, flow.state)
    if updated:
        return updated
    raise HTTPException(status_code=404, detail="Flow not found")


@router.delete("/{id}", description="Delete a flow by ID", response_model=FlowOut)
def delete_flow(id: int, db: Session = Depends(get_db)):
    deleted = delete_flow_by_id(db, id)
    if deleted:
        return deleted
    raise HTTPException(status_code=404, detail="Flow not found")

##################
## Code Generation
##################

@router.post("/generate/code", description="Generate flow code by submitting the canvas graph")
def generate_flow_code(flow: FlowPayload):
    print(flow)
    codegen = CodeGenerator()
    try:
        code = codegen.generate(flow)
    except Exception as e:
        print(f"Code generation error: {e}")
        raise HTTPException(status_code=500, detail=f"{e}")
    return {"code": code}


@router.post("/{id}/code", description="Generate flow code from saved flow")
def generate_flow_code(id: int):
    return {"Hello": "World"}


@router.post("/{id}/code", description="Generate flow code from saved flow")
def generate_flow_code(id: int):
    return {"Hello": "World"}


@router.post("/{id}/test", description="Test a flow")
def test_flow(id: int):
    return {"Hello": "World"}
